### ISSUE TITLE
Verifying peer X.509 certificate: failed

### ISSUE THREAD

**Issue by bmoraine** [ARMmbed/mbedtls#421]
Hello,

When evaluating mbedtls, I notice a strange behavior running  ssl_client1 with ssl_server example programs that share a correct set of certificates (in my understanding).

Indeed, on client side during hanshaking, the X.509 certificate verification failed although on sever side no error is reported

For information, please find below the output of ssl_client1 when I meet the issue :

```
  . Seeding the random number generator... ok
  . Loading the CA root certificate ... ok (0 skipped)
  . Connecting to tcp/localhost/4433... ok
  . Setting up the SSL/TLS structure... ok
  . Performing the SSL/TLS handshake.../home/bmoraine/Desktop/mbed/mbedtls-2.2.0/library/ssl_tls.c:4400: x509_verify_cert() returned -9984 (-0x2700)
 ok
  . Verifying peer X.509 certificate... failed
  ! The certificate Common Name (CN) does not match with the expected CN

  > Write to server: 18 bytes written

GET / HTTP/1.0

  < Read from server: 150 bytes read

HTTP/1.0 200 OK
Content-Type: text/html

<h2>mbed TLS Test Server</h2>
<p>Successful connection using: TLS-ECDHE-RSA-WITH-AES-256-GCM-SHA384</p>
/home/bmoraine/Desktop/mbed/mbedtls-2.2.0/library/ssl_tls.c:6509: mbedtls_ssl_read_record() returned -30848 (-0x7880)
Last error was: -30848 - SSL - The peer notified us that the connection is going to be closed
```

Regarding ssl_server output no error is displayed :

```
  . Loading the server cert. and key... ok
  . Bind on https://localhost:4433/ ... ok
  . Seeding the random number generator... ok
  . Setting up the SSL data.... ok
  . Waiting for a remote connection ... ok
  . Performing the SSL/TLS handshake... ok
  < Read from client: 18 bytes read

GET / HTTP/1.0

  > Write to client: 150 bytes written

HTTP/1.0 200 OK
Content-Type: text/html

<h2>mbed TLS Test Server</h2>
<p>Successful connection using: TLS-ECDHE-RSA-WITH-AES-256-GCM-SHA384</p>

  . Closing the connection... ok
  . Waiting for a remote connection ...
```

I reproduced the same issue using unmodified ssl_client1 and ssl_server example programs on Ubuntu 14.10 for the following versions : v2.2.1, v2.2.0 ,v2.1.4 , v1.3.16, but not with v1.2.19.

Is there a regression in ssl_client1 example, in the used certificates or in the library itself?

Or maybe I am missing something or did something wrong ....

Thank you for your help

**Comment by bmoraine** [comment]
It seems I fix the issue by replacing hostname parameter in the call of mbedtls_ssl_set_hostname() on line 180

I replace : 
~~~
  if( ( ret = mbedtls_ssl_set_hostname( &ssl, "mbed TLS Server 1" ) ) != 0 )
   {
        mbedtls_printf( " failed\n  ! mbedtls_ssl_set_hostname returned %d\n\n", ret );
        goto exit;
    }
~~~
by 
~~~
  if( ( ret = mbedtls_ssl_set_hostname( &pms->ssl, SERVER_NAME ) ) != 0 )
    {
        mbedtls_printf( " failed\n  ! mbedtls_ssl_set_hostname returned %d\n\n", ret );
        goto exit;
    }
~~~

For information, SERVER_NAME is defined on line 63
~~~
#define SERVER_NAME "localhost"
~~~
and is used previously used by mbedtls_net_connect()on line 141

**Comment by ciarmcom** [comment]
ARM Internal Ref: IOTSSL-636

**Comment by sbutcher-arm** [comment]
Can I ask where you got your certificates from and if you've tested them independently to check if they're valid?

**Comment by bmoraine** [comment]
i did not modify the source code of the  ssl_client1 & ssl_server examples.

The used certificates are the default ones.

These certificates are not files but defined in library/certs.c

Did I answer to your question?

**Comment by sbutcher-arm** [comment]
Fixed in PR #548, and merged with `development`, `mbedtls-2.4`, and `mbedtls-2.1`.
